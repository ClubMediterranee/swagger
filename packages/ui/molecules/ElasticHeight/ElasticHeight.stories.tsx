import { StoryFn, Meta } from '@storybook/react';
import { useState } from 'react';

import { Button } from './Buttons/Button';

import { ElasticHeight } from './ElasticHeight';

export default {
  title: 'UI-Library/Component/Molecules/ElasticHeight',
  component: ElasticHeight,
} as Meta<typeof ElasticHeight>;

const Template: StoryFn<typeof ElasticHeight> = () => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <div>
      <Button onClick={() => setIsOpen(!isOpen)} label="Toggle" />
      <ElasticHeight>
        <div className={isOpen ? '' : 'h-0'}>
          <table>
            <caption>
              <i>Being Funny in a Foreign Language</i> track listing
            </caption>
            <tbody>
              <tr>
                <th scope="col">
                  <abbr title="Number">No.</abbr>
                </th>
                <th scope="col" style={{ width: '50%' }}>
                  Title
                </th>
                <th scope="col" style={{ width: '50%' }}>
                  Writer(s)
                </th>
                <th scope="col">Length</th>
              </tr>
              <tr>
                <th id="track1" scope="row">
                  1.
                </th>
                <td>"The 1975"</td>
                <td>
                  <div>
                    <ul>
                      <li>Matthew Healy</li>
                      <li>George Daniel</li>
                      <li>BJ Burton</li>
                      <li>Tommy King</li>
                    </ul>
                  </div>
                </td>
                <td>4:10</td>
              </tr>
              <tr>
                <th id="track2" scope="row">
                  2.
                </th>
                <td>"Happiness"</td>
                <td>
                  <div>
                    <ul>
                      <li>Healy</li>
                      <li>Daniel</li>
                      <li>DJ Sabrina the Teenage DJ</li>
                    </ul>
                  </div>
                </td>
                <td>5:03</td>
              </tr>
              <tr>
                <th id="track3" scope="row">
                  3.
                </th>
                <td>"Looking for Somebody (To Love)"</td>
                <td>
                  <div>
                    <ul>
                      <li>Healy</li>
                      <li>Daniel</li>
                      <li>Ilsey Juber</li>
                      <li>Jamie Squire</li>
                    </ul>
                  </div>
                </td>
                <td>2:58</td>
              </tr>
              <tr>
                <th id="track4" scope="row">
                  4.
                </th>
                <td>"Part of the Band"</td>
                <td>
                  <div>
                    <ul>
                      <li>Healy</li>
                      <li>Daniel</li>
                      <li>Squire</li>
                    </ul>
                  </div>
                </td>
                <td>4:20</td>
              </tr>
              <tr>
                <th id="track5" scope="row">
                  5.
                </th>
                <td>"Oh Caroline"</td>
                <td>
                  <div>
                    <ul>
                      <li>Healy</li>
                      <li>Daniel</li>
                      <li>Juber</li>
                      <li>Squire</li>
                      <li>Benjamin Francis Leftwich</li>
                      <li>Jimmy Hogarth</li>
                    </ul>
                  </div>
                </td>
                <td>3:32</td>
              </tr>
              <tr>
                <th id="track6" scope="row">
                  6.
                </th>
                <td>"I'm in Love with You"</td>
                <td>
                  <div>
                    <ul>
                      <li>Healy</li>
                      <li>Daniel</li>
                      <li>Adam Hann</li>
                    </ul>
                  </div>
                </td>
                <td>4:22</td>
              </tr>
              <tr>
                <th id="track7" scope="row">
                  7.
                </th>
                <td>"All I Need to Hear"</td>
                <td>
                  <div>
                    <ul>
                      <li>Healy</li>
                      <li>Daniel</li>
                      <li>Squire</li>
                    </ul>
                  </div>
                </td>
                <td>3:30</td>
              </tr>
              <tr>
                <th id="track8" scope="row">
                  8.
                </th>
                <td>"Wintering"</td>
                <td>
                  <div>
                    <ul>
                      <li>Healy</li>
                      <li>Daniel</li>
                      <li>Jacob Bugden</li>
                    </ul>
                  </div>
                </td>
                <td>2:45</td>
              </tr>
              <tr>
                <th id="track9" scope="row">
                  9.
                </th>
                <td>"Human Too"</td>
                <td>
                  <div>
                    <ul>
                      <li>Healy</li>
                      <li>Daniel</li>
                      <li>Squire</li>
                      <li>Leftwich</li>
                      <li>Hogarth</li>
                    </ul>
                  </div>
                </td>
                <td>3:44</td>
              </tr>
              <tr>
                <th id="track10" scope="row">
                  10.
                </th>
                <td>"About You"</td>
                <td>
                  <div>
                    <ul>
                      <li>Healy</li>
                      <li>Daniel</li>
                    </ul>
                  </div>
                </td>
                <td>5:26</td>
              </tr>
              <tr>
                <th id="track11" scope="row">
                  11.
                </th>
                <td>"When We Are Together"</td>
                <td>
                  <div>
                    <ul>
                      <li>Healy</li>
                      <li>Daniel</li>
                      <li>Rob Milton</li>
                    </ul>
                  </div>
                </td>
                <td>3:36</td>
              </tr>
              <tr>
                <th colSpan={3} scope="row">
                  <span>Total length:</span>
                </th>
                <td>43:26</td>
              </tr>
            </tbody>
          </table>
        </div>
      </ElasticHeight>
    </div>
  );
};

export const Default = Template.bind({});
