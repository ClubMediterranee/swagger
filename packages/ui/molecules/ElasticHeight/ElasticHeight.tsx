import { useSpring, animated, useResize } from '@react-spring/web';
import classnames from 'classnames';
import { FunctionComponent, MutableRefObject, PropsWithChildren, useEffect, useRef } from 'react';

/**
 * This is not a Collapse component, it's a component that animates *to* the height of its children.
 * It does the *absolute* same but you Bring Your Own Heightâ„¢ by applying a class to the children.
 * e.g.: <ElasticHeight className={isOpen ? "" : "h-0"}><div>...</div></ElasticHeight>
 * ID is there for accessibility purposes.
 *
 * This is in no way affiliated with Amazon ElasticSearch.
 */

interface ElasticHeightProps {
  /**
   * Id of the element
   */
  id?: string;
  /**
   * Class of the element (please provide something that sets the height)
   */
  className?: string;
  /**
   * Class of the container (to avoid CLS, mainly)
   */
  containerClassName?: string;
  /**
   * Role of the element
   */
  role?: string;
}

const linearInterpolation = (x: number) => x;

export const ElasticHeight: FunctionComponent<PropsWithChildren<ElasticHeightProps>> = ({
  children,
  id,
  className,
  role,
  containerClassName,
}) => {
  const ref = useRef<HTMLDivElement>(null) as MutableRefObject<HTMLDivElement>;
  const { height } = useResize({ container: ref });

  const [styles, api] = useSpring(() => ({
    from: {
      height: height.to(linearInterpolation),
    },
    immediate: true,
  }));

  useEffect(() => {
    api.start({
      to: {
        height: height.to(linearInterpolation),
      },
      config: {
        tension: 170,
        friction: 26,
      },
    });
  }, [api, height]);

  return (
    <animated.div
      role={role}
      className={classnames('overflow-hidden', containerClassName)}
      style={{ ...styles }}
      id={id}
    >
      <div className={className} ref={ref}>
        {children}
      </div>
    </animated.div>
  );
};
