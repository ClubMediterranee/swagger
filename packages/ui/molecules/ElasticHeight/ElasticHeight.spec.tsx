import { act, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useState } from 'react';

import { ElasticHeight } from './ElasticHeight';

describe('<ElasticHeight />', () => {
  it('keeps the content in the DOM even when closed', () => {
    render(
      <ElasticHeight>
        <p>Awesome content</p>
      </ElasticHeight>,
    );

    const content = screen.getByText('Awesome content');

    expect(content).toBeInTheDocument();
  });

  it('shows the content when open', async () => {
    const TestComponent = () => {
      const [isOpen, setIsOpen] = useState(false);
      return (
        <div role="presentation">
          {/* eslint-disable-next-line jest/no-conditional-in-test */}
          <ElasticHeight className={isOpen ? '' : 'h-0'}>
            <p>Awesome content</p>
          </ElasticHeight>
          <button onClick={() => setIsOpen(true)}>open</button>
        </div>
      );
    };

    render(<TestComponent />);
    const component = screen.getByRole('presentation');
    const button = screen.getByRole('button');
    await act(() => userEvent.click(button));

    expect(component).toMatchSnapshot('open elasticheight');
  });
});
