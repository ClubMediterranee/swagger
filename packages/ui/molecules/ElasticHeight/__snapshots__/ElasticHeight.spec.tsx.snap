// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ElasticHeight /> shows the content when open: open elasticheight 1`] = `
<div
  role="presentation"
>
  <div
    class="overflow-hidden"
    style="height: 0px;"
  >
    <div
      class=""
    >
      <p>
        Awesome content
      </p>
    </div>
  </div>
  <button>
    open
  </button>
</div>
`;
